File: app/src/main/java/com/adamhurwitz/retrorecycler/DependencyInjection/DataModule.java
package com.adamhurwitz.retrorecycler.DependencyInjection;

import android.app.Application;

import com.adamhurwitz.retrorecycler.Network.Repository;
import com.adamhurwitz.retrorecycler.Network.Service;

import javax.inject.Singleton;

import dagger.Module;
import dagger.Provides;
import okhttp3.OkHttpClient;
import retrofit2.GsonConverterFactory;
import retrofit2.Retrofit;
import retrofit2.RxJavaCallAdapterFactory;

/**
 * Created by ahurwitz on 1/24/17.
 */

@Module
public class DataModule {

    public final static String BASE_URL = "https://www.udacity.com";

    Application application;

    public DataModule(Application application) {
        this.application = application;
    }

    @Provides
    @Singleton
    public OkHttpClient provideHttpClient() {
        return new OkHttpClient().newBuilder().build();
    }

    @Provides
    @Singleton
    Retrofit provideRetrofit(OkHttpClient okHttpClient) {
        Retrofit retrofit = new Retrofit.Builder()
                .addConverterFactory(GsonConverterFactory.create())
                .baseUrl(BASE_URL)
                .client(okHttpClient)
                //converts Retrofit response into Observable
                .addCallAdapterFactory(RxJavaCallAdapterFactory.create())
                .build();
        return retrofit;
    }

    @Provides
    public Service provideItemService(Retrofit retrofit) {
        return retrofit.create(Service.class);
    }

    @Provides
    public Repository provideRepository(Service service) {
        return new Repository(service);
    }


}

File: app/src/main/java/io/github/adamshurwitz/retrorecycler/dependencyinjection/DataModule.kt
package io.github.adamshurwitz.retrorecycler.dependencyinjection

import android.app.Application
import dagger.Module
import dagger.Provides
import io.github.adamshurwitz.retrorecycler.network.Repository
import io.github.adamshurwitz.retrorecycler.network.Service
import okhttp3.OkHttpClient
import retrofit2.GsonConverterFactory
import retrofit2.Retrofit
import retrofit2.RxJavaCallAdapterFactory
import javax.inject.Singleton

/**
 * Created by ahurwitz on 1/24/17.
 */

@Module
class DataModule(internal var application: Application) {

    companion object {
        val BASE_URL = "https://www.udacity.com"
    }

    @Provides
    @Singleton
    fun provideHttpClient(): OkHttpClient {
        return OkHttpClient().newBuilder().build()
    }

    @Provides
    @Singleton
    fun provideRetrofit(okHttpClient: OkHttpClient): Retrofit {
        val retrofit = Retrofit.Builder()
                .addConverterFactory(GsonConverterFactory.create())
                .baseUrl(BASE_URL)
                .client(okHttpClient)
                .addCallAdapterFactory(RxJavaCallAdapterFactory.create())
                .build()
        return retrofit
    }

    @Provides
    fun provideItemService(retrofit: Retrofit): Service {
        return retrofit.create(Service::class.java)
    }

    @Provides
    fun provideRepository(service: Service): Repository {
        return Repository(service)
    }


}